File: app/src/main/java/org/allatra/radio/MainActivity.java
package org.allatra.radio;

import android.Manifest;
import android.annotation.SuppressLint;
import android.content.ComponentName;
import android.content.ContentResolver;
import android.content.Context;
import android.content.DialogInterface;
import android.content.Intent;
import android.content.ServiceConnection;
import android.content.pm.PackageManager;
import android.content.res.Configuration;
import android.database.Cursor;
import android.media.AudioManager;
import android.net.Uri;
import android.os.Build;
import android.os.Bundle;
import android.os.IBinder;
import android.provider.MediaStore;
import android.support.annotation.NonNull;
import android.support.v4.app.ActivityCompat;
import android.support.v4.content.ContextCompat;
import android.support.v7.app.AlertDialog;
import android.support.v7.app.AppCompatActivity;
import android.telephony.PhoneStateListener;
import android.telephony.TelephonyManager;
import android.view.View;
import android.widget.CheckBox;
import android.widget.ImageButton;
import android.widget.ImageView;
import android.widget.Toast;

import java.io.IOException;
import java.util.ArrayList;

import wseemann.media.FFmpegMediaPlayer;

public class MainActivity extends AppCompatActivity
        implements
        View.OnClickListener,
        AudioManager.OnAudioFocusChangeListener {

    private static final int MY_PERMISSIONS_REQUEST_READ_PHONE_STATE = 1;
    private ImageButton buttonPlay;
    private ImageButton buttonStopPlay;
    private CheckBox checkRadio;
    private CheckBox checkClassic;
    public static final String RADIO_STREAM = "http://94.23.36.117:5551/stream";
    public static final String CLASSIC_STREAM = "http://eu6.fastcast4u.com:5551/classic";
    public static final String Broadcast_PLAY_NEW_AUDIO = "org.allatra.radio.audioplayer.PlayNewAudio";
    private FFmpegMediaPlayer fFmpegMediaPlayer;

    private AudioPlayerService audioPlayerService;
    boolean serviceBound = false;
    private ArrayList<Audio> audioList;

    //Handle incoming phone calls
    private boolean ongoingCall = false;
    private int permissionCheck;
    private AlertDialog.Builder dialogFragment;

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        initializeUIElements();
        if (permissionCheck == 1) callStateListener();
        else permission();

//        loadAudio();
        //play the first audio in the ArrayList
//        playAudio(audioList.get(0).getData());

    }


    @Override
    protected void onStart() {
        super.onStart();
        startService();
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
        if (serviceBound) {
            unbindService(serviceConnection);
            //service is active
            audioPlayerService.stopSelf();
        }
//        releaseMP();
    }

    @Override
    protected void onPause() {
//        fFmpegMediaPlayer.stop();
//        fFmpegMediaPlayer.release();
        super.onPause();
    }

    @Override
    protected void onResume() {
//        if (checkRadio.isChecked())
//            startPlaying();
//            fFmpegMediaPlayer = new FFmpegMediaPlayer();
//        fFmpegMediaPlayer.setLooping(false);
//        fFmpegMediaPlayer.start();
        super.onResume();
    }

    @Override
    public void onSaveInstanceState(Bundle savedInstanceState) {
        savedInstanceState.putBoolean("ServiceState", serviceBound);
        super.onSaveInstanceState(savedInstanceState);
    }

    @Override
    public void onRestoreInstanceState(Bundle savedInstanceState) {
        super.onRestoreInstanceState(savedInstanceState);
        serviceBound = savedInstanceState.getBoolean("ServiceState");
    }

    @Override
    public void onRequestPermissionsResult(int requestCode,
                                           @NonNull String permissions[], @NonNull int[] grantResults) {
        switch (requestCode) {
            case MY_PERMISSIONS_REQUEST_READ_PHONE_STATE: {
                // If request is cancelled, the result arrays are empty.
                if (grantResults.length > 0
                        && grantResults[0] == PackageManager.PERMISSION_GRANTED) {
                    // permission was granted, yay! Do the
                    // contacts-related task you need to do.
                    callStateListener();
                } else {
                    // permission denied, boo! Disable the
                    // functionality that depends on this permission.
                    dialogFragment.create();
                }
            }

            // other 'case' lines to check for other
            // permissions this app might request
        }
    }

    private void permission() {
        // Here, thisActivity is the current activity
        if (ContextCompat.checkSelfPermission(this,
                Manifest.permission.READ_PHONE_STATE)
                != PackageManager.PERMISSION_GRANTED) {

            // Should we show an explanation?
            if (ActivityCompat.shouldShowRequestPermissionRationale(this,
                    Manifest.permission.READ_PHONE_STATE)) {

                // Show an explanation to the user *asynchronously* -- don't block
                // this thread waiting for the user's response! After the user
                // sees the explanation, try again to request the permission.

            } else {

                // No explanation needed, we can request the permission.

                ActivityCompat.requestPermissions(this,
                        new String[]{Manifest.permission.READ_PHONE_STATE},
                        MY_PERMISSIONS_REQUEST_READ_PHONE_STATE);

                // MY_PERMISSIONS_REQUEST_READ_CONTACTS is an
                // app-defined int constant. The callback method gets the
                // result of the request.
            }
        }
    }

    @SuppressLint("ShowToast")
    private void initializeUIElements() {
        buttonPlay = findViewById(R.id.buttonPlay);
        buttonPlay.setOnClickListener(this);
        buttonPlay.setEnabled(true);

        buttonStopPlay = findViewById(R.id.buttonStopPlay);
        buttonStopPlay.setEnabled(false);
        buttonStopPlay.setVisibility(View.INVISIBLE);
        buttonStopPlay.setOnClickListener(this);

        checkRadio = findViewById(R.id.chb_radio);
        checkClassic = findViewById(R.id.chb_classic);
        checkRadio.setChecked(true);

        permissionCheck = ContextCompat.checkSelfPermission(this,
                Manifest.permission.READ_PHONE_STATE);

        dialogFragment = new AlertDialog.Builder(this)
                .setTitle(R.string.attantion)
                .setMessage(R.string.message_text)
                .setNeutralButton(R.string.yes, new DialogInterface.OnClickListener() {
                    public void onClick(DialogInterface dialog, int which) {
                        dialog.dismiss();
                        finish();
                    }
                });

        if (this.getResources().getConfiguration().orientation == Configuration.ORIENTATION_PORTRAIT) {
            // Portrait Mode
            ImageView background = findViewById(R.id.iv_portrait);
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.JELLY_BEAN) {
                background.setBackground(getResources().getDrawable(R.drawable.back));
            } else {
                background.setImageDrawable(getResources().getDrawable(R.drawable.back));
            }
        } else {
            // Landscape Mode
            ImageView background = findViewById(R.id.iv_land);
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.JELLY_BEAN) {
                background.setBackground(getResources().getDrawable(R.drawable.back_land));
            } else {
                background.setImageDrawable(getResources().getDrawable(R.drawable.back_land));
            }
        }

    }

    public void onClick(View v) {
        if (serviceBound)
            if (v == buttonPlay) {
                if (checkRadio.isChecked()) {
                    audioPlayerService.startPlayRadio();
                } else {
                    audioPlayerService.startPlayClassic();
                }
                buttonPlay.setVisibility(View.INVISIBLE);
                buttonStopPlay.setVisibility(View.VISIBLE);
                buttonStopPlay.setEnabled(true);
                buttonPlay.setEnabled(false);

                checkRadio.setEnabled(false);
                checkClassic.setEnabled(false);
//            startPlaying();
            } else if (v == buttonStopPlay) {
                audioPlayerService.stopMedia();

                buttonStopPlay.setVisibility(View.INVISIBLE);
                buttonPlay.setEnabled(true);
                buttonPlay.setVisibility(View.VISIBLE);
                buttonStopPlay.setEnabled(false);
                checkClassic.setEnabled(true);
                checkRadio.setEnabled(true);
//            stopPlaying();
            }
    }

    private void startPlaying() {
        fFmpegMediaPlayer = new FFmpegMediaPlayer();
        fFmpegMediaPlayer.setOnPreparedListener(new FFmpegMediaPlayer.OnPreparedListener() {

            @Override
            public void onPrepared(FFmpegMediaPlayer mp) {
                mp.start();
            }
        });

        fFmpegMediaPlayer.setOnErrorListener(new FFmpegMediaPlayer.OnErrorListener() {

            @Override
            public boolean onError(FFmpegMediaPlayer mp, int what, int extra) {
                mp.release();
                return false;
            }
        });

        try {
            if (checkRadio.isChecked()) {
                fFmpegMediaPlayer.setDataSource(RADIO_STREAM);
            } else {
                fFmpegMediaPlayer.setDataSource(CLASSIC_STREAM);
            }
            fFmpegMediaPlayer.prepareAsync();
        } catch (IllegalArgumentException e) {
            e.printStackTrace();
        } catch (SecurityException e) {
            e.printStackTrace();
        } catch (IllegalStateException e) {
            e.printStackTrace();
        } catch (IOException e) {
            e.printStackTrace();
        }

        buttonPlay.setVisibility(View.INVISIBLE);
        buttonStopPlay.setVisibility(View.VISIBLE);
        buttonStopPlay.setEnabled(true);
        buttonPlay.setEnabled(false);

        checkRadio.setEnabled(false);
        checkClassic.setEnabled(false);
    }

    private void stopPlaying() {
        if (fFmpegMediaPlayer.isPlaying())
            fFmpegMediaPlayer.stop();

        buttonStopPlay.setVisibility(View.INVISIBLE);
        buttonPlay.setEnabled(true);
        buttonPlay.setVisibility(View.VISIBLE);
        buttonStopPlay.setEnabled(false);
        checkClassic.setEnabled(true);
        checkRadio.setEnabled(true);

    }

    private void releaseMP() {
        if (fFmpegMediaPlayer != null) {
            try {
                fFmpegMediaPlayer.release();
                fFmpegMediaPlayer = null;
            } catch (Exception e) {
                e.printStackTrace();
            }
        }

    }

    @Override
    public void onAudioFocusChange(int focusChange) {
        switch (focusChange) {
            case AudioManager.AUDIOFOCUS_GAIN:
                // resume playback
                if (fFmpegMediaPlayer == null) startPlaying();
                else if (!fFmpegMediaPlayer.isPlaying()) fFmpegMediaPlayer.start();
                fFmpegMediaPlayer.setVolume(1.0f, 1.0f);
                break;

            case AudioManager.AUDIOFOCUS_LOSS:
                // Lost focus for an unbounded amount of time: stop playback and release media player
                if (fFmpegMediaPlayer.isPlaying()) fFmpegMediaPlayer.stop();
                fFmpegMediaPlayer.release();
                fFmpegMediaPlayer = null;
                break;

            case AudioManager.AUDIOFOCUS_LOSS_TRANSIENT:
                // Lost focus for a short time, but we have to stop
                // playback. We don't release the media player because playback
                // is likely to resume
                if (fFmpegMediaPlayer.isPlaying()) fFmpegMediaPlayer.pause();
                break;

            case AudioManager.AUDIOFOCUS_LOSS_TRANSIENT_CAN_DUCK:
                // Lost focus for a short time, but it's ok to keep playing
                // at an attenuated level
                if (fFmpegMediaPlayer.isPlaying()) fFmpegMediaPlayer.pause();
                break;
        }
    }

    public void onCheckBoxClicked(View view) {
        // Is the view now checked?
        boolean checked = ((CheckBox) view).isChecked();

        // Check which checkbox was clicked
        switch (view.getId()) {
            case R.id.chb_radio:
                if (checked) {
                    checkClassic.setChecked(false);
                }
                break;
            case R.id.chb_classic:
                if (checked) {
                    checkRadio.setChecked(false);
                }
                break;
        }
    }

    private void callStateListener() {
        // Get the telephony manager
        TelephonyManager telephonyManager = (TelephonyManager) getSystemService(Context.TELEPHONY_SERVICE);
        //Starting listening for PhoneState changes
        PhoneStateListener phoneStateListener = new PhoneStateListener() {
            @Override
            public void onCallStateChanged(int state, String incomingNumber) {
                switch (state) {
                    //if at least one call exists or the phone is ringing
                    //pause the MediaPlayer
                    case TelephonyManager.CALL_STATE_OFFHOOK:
                    case TelephonyManager.CALL_STATE_RINGING:
                        if (fFmpegMediaPlayer != null) {
                            fFmpegMediaPlayer.pause();
                            ongoingCall = true;
                        }
                        break;
                    case TelephonyManager.CALL_STATE_IDLE:
                        // Phone idle. Start playing.
                        if (fFmpegMediaPlayer != null) {
                            if (ongoingCall) {
                                ongoingCall = false;
                                fFmpegMediaPlayer.start();
                            }
                        }
                        break;
                }
            }
        };
        // Register the listener with the telephony manager
        // Listen for changes to the device call state.
        telephonyManager.listen(phoneStateListener,
                PhoneStateListener.LISTEN_CALL_STATE);
    }

    private void playAudio(int audioIndex) {
        //Check is service is active
        if (!serviceBound) {
            //Store Serializable audioList to SharedPreferences
            StorageUtil storage = new StorageUtil(getApplicationContext());
            storage.storeAudio(audioList);
            storage.storeAudioIndex(audioIndex);

            Intent playerIntent = new Intent(this, AudioPlayerService.class);
            startService(playerIntent);
            bindService(playerIntent, serviceConnection, Context.BIND_AUTO_CREATE);
        } else {
            //Store the new audioIndex to SharedPreferences
            StorageUtil storage = new StorageUtil(getApplicationContext());
            storage.storeAudioIndex(audioIndex);

            //Service is active
            //Send a broadcast to the service -> PLAY_NEW_AUDIO
            Intent broadcastIntent = new Intent(Broadcast_PLAY_NEW_AUDIO);
            sendBroadcast(broadcastIntent);
        }
    }

    private void loadAudio() {
        ContentResolver contentResolver = getContentResolver();

        Uri uri = MediaStore.Audio.Media.EXTERNAL_CONTENT_URI;
        String selection = MediaStore.Audio.Media.IS_MUSIC + "!= 0";
        String sortOrder = MediaStore.Audio.Media.TITLE + " ASC";
        Cursor cursor = contentResolver.query(uri, null, selection, null, sortOrder);

        if (cursor != null && cursor.getCount() > 0) {
            audioList = new ArrayList<>();
            while (cursor.moveToNext()) {
                String data = cursor.getString(cursor.getColumnIndex(MediaStore.Audio.Media.DATA));
                String title = cursor.getString(cursor.getColumnIndex(MediaStore.Audio.Media.TITLE));
                String album = cursor.getString(cursor.getColumnIndex(MediaStore.Audio.Media.ALBUM));
                String artist = cursor.getString(cursor.getColumnIndex(MediaStore.Audio.Media.ARTIST));

                // Save to audioList
                audioList.add(new Audio(data, title, album, artist));
            }
        }
        cursor.close();
    }

    //Binding this Client to the AudioPlayer Service
    private ServiceConnection serviceConnection = new ServiceConnection() {
        @Override
        public void onServiceConnected(ComponentName name, IBinder service) {
            // We've bound to LocalService, cast the IBinder and get LocalService instance
            AudioPlayerService.LocalBinder binder = (AudioPlayerService.LocalBinder) service;
            audioPlayerService = binder.getService();
            serviceBound = true;

            Toast.makeText(MainActivity.this, "Service Bound", Toast.LENGTH_SHORT).show();
        }

        @Override
        public void onServiceDisconnected(ComponentName name) {
            serviceBound = false;
        }
    };

    private void startService() {
        Intent playerIntent = new Intent(this, AudioPlayerService.class);
        startService(playerIntent);
        bindService(playerIntent, serviceConnection, Context.BIND_AUTO_CREATE);
    }
}

File: app/src/main/java/org/allatra/radio/MainActivity.kt
package org.allatra.radio

import android.Manifest
import android.content.Context
import android.content.pm.PackageManager
import android.content.res.Configuration
import android.media.AudioManager
import android.media.AudioManager.OnAudioFocusChangeListener
import android.net.Uri
import android.os.Bundle
import android.telephony.PhoneStateListener
import android.telephony.TelephonyManager
import android.view.View
import android.widget.CheckBox
import androidx.appcompat.app.AlertDialog
import androidx.appcompat.app.AppCompatActivity
import androidx.core.app.ActivityCompat
import androidx.core.content.ContextCompat
import com.google.android.exoplayer2.SimpleExoPlayer
import com.google.android.exoplayer2.source.MediaSource
import com.google.android.exoplayer2.source.ProgressiveMediaSource
import com.google.android.exoplayer2.upstream.DataSource
import com.google.android.exoplayer2.upstream.DefaultDataSourceFactory
import com.google.android.exoplayer2.util.Util
import kotlinx.android.synthetic.main.activity_main.*


class MainActivity : AppCompatActivity(), View.OnClickListener, OnAudioFocusChangeListener {
    private var uri: Uri = Uri.parse(RADIO_STREAM)
    private var audioPlayerService: AudioPlayerService? = null
    var serviceBound = false
    private var playWhenReady = true


    //Handle incoming phone calls
    private val ongoingCall = false
    private var permissionCheck = 0
    private var dialogFragment: AlertDialog.Builder? = null
    private var player: SimpleExoPlayer? = null
    public override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)
        initializeUIElements()
        if (permissionCheck == 1) callStateListener() else permission()
    }

    private fun initializePlayer() {
        player = SimpleExoPlayer.Builder(this).build()
    }

    private fun buildMediaSource(uri: Uri): MediaSource? {
        val dataSourceFactory: DataSource.Factory = DefaultDataSourceFactory(this, "exoplayer-allatraradio")
        return ProgressiveMediaSource.Factory(dataSourceFactory)
                .createMediaSource(uri)
    }

    override fun onStart() {
        super.onStart()
        if (Util.SDK_INT >= 24) {
            initializePlayer()
        }
    }

    override fun onResume() {
        super.onResume()
        if (Util.SDK_INT < 24 || player == null) {
            initializePlayer()
        }
    }

    override fun onPause() {
        super.onPause()
        if (Util.SDK_INT < 24) {
            releasePlayer()
        }
    }

    override fun onStop() {
        super.onStop()
        if (Util.SDK_INT >= 24) {
            releasePlayer()
        }
    }

    override fun onDestroy() {
        super.onDestroy()
        player!!.release()
    }

    private fun releasePlayer() {
        playWhenReady = player!!.playWhenReady
        player!!.release()
        player = null
    }

    public override fun onSaveInstanceState(savedInstanceState: Bundle) {
        savedInstanceState.putBoolean("ServiceAllatRaRadioState", serviceBound)
        super.onSaveInstanceState(savedInstanceState)
    }

    public override fun onRestoreInstanceState(savedInstanceState: Bundle) {
        super.onRestoreInstanceState(savedInstanceState)
        serviceBound = savedInstanceState.getBoolean("ServiceAllatRaRadioState")
    }

    override fun onRequestPermissionsResult(requestCode: Int,
                                            permissions: Array<String>, grantResults: IntArray) {
        when (requestCode) {
            MY_PERMISSIONS_REQUEST_READ_PHONE_STATE -> {

                // If request is cancelled, the result arrays are empty.
                if (grantResults.size > 0
                        && grantResults[0] == PackageManager.PERMISSION_GRANTED) {
                    // permission was granted, yay! Do the
                    // contacts-related task you need to do.
                    callStateListener()
                } else {
                    // permission denied, boo! Disable the
                    // functionality that depends on this permission.
                    dialogFragment!!.create()
                }
            }
        }
    }

    private fun permission() {
        // Here, thisActivity is the current activity
        if (ContextCompat.checkSelfPermission(this,
                        Manifest.permission.READ_PHONE_STATE)
                != PackageManager.PERMISSION_GRANTED) {

            // Should we show an explanation?
            if (ActivityCompat.shouldShowRequestPermissionRationale(this,
                            Manifest.permission.READ_PHONE_STATE)) {

                // Show an explanation to the user *asynchronously* -- don't block
                // this thread waiting for the user's response! After the user
                // sees the explanation, try again to request the permission.
            } else {

                // No explanation needed, we can request the permission.
                ActivityCompat.requestPermissions(this, arrayOf(Manifest.permission.READ_PHONE_STATE),
                        MY_PERMISSIONS_REQUEST_READ_PHONE_STATE)

                // MY_PERMISSIONS_REQUEST_READ_CONTACTS is an
                // app-defined int constant. The callback method gets the
                // result of the request.
            }
        }
    }

    private fun initializeUIElements() {
        buttonPlay.setOnClickListener(this)
        buttonPlay.isEnabled = true
        buttonStopPlay.isEnabled = false
        buttonStopPlay.visibility = View.INVISIBLE
        buttonStopPlay.setOnClickListener(this)
        chb_radio.isChecked = true

        permissionCheck = ContextCompat.checkSelfPermission(this,
                Manifest.permission.READ_PHONE_STATE)

        dialogFragment = AlertDialog.Builder(this)
                .setTitle(R.string.attantion)
                .setMessage(R.string.message_text)
                .setNeutralButton(R.string.yes) { dialog, which ->
                    dialog.dismiss()
                    finish()
                }

        if (this.resources.configuration.orientation == Configuration.ORIENTATION_PORTRAIT) {
            // Portrait Mode
            iv_portrait.background = ContextCompat.getDrawable(this, R.drawable.back)
        } else {
            // Landscape Mode
            iv_land.background = ContextCompat.getDrawable(this, R.drawable.back_land)
        }
    }

    override fun onClick(v: View) {
//        if (serviceBound) {
        if (v === buttonPlay) {
            if (chb_radio!!.isChecked) {
//                    audioPlayerService!!.startPlayRadio()
                uri = Uri.parse(RADIO_STREAM)
            } else {
                uri = Uri.parse(CLASSIC_STREAM)
//                    audioPlayerService!!.startPlayClassic()
            }
            buttonPlay!!.visibility = View.INVISIBLE
            buttonStopPlay!!.visibility = View.VISIBLE
            buttonStopPlay!!.isEnabled = true
            buttonPlay!!.isEnabled = false
            chb_radio!!.isEnabled = false
            chb_classic!!.isEnabled = false
            startPlaying()
        } else if (v === buttonStopPlay) {
//                audioPlayerService!!.stopMedia()
            buttonStopPlay!!.visibility = View.INVISIBLE
            buttonPlay!!.isEnabled = true
            buttonPlay!!.visibility = View.VISIBLE
            buttonStopPlay!!.isEnabled = false
            chb_classic!!.isEnabled = true
            chb_radio!!.isEnabled = true
            stopPlaying()
        }
//        }
    }

    private fun startPlaying() {
        player.run {
            val mediaSource = buildMediaSource(uri)
            mediaSource?.let { player!!.prepare(it, false, false) }
            player!!.playWhenReady = playWhenReady
        }

        buttonPlay!!.visibility = View.INVISIBLE
        buttonStopPlay!!.visibility = View.VISIBLE
        buttonStopPlay!!.isEnabled = true
        buttonPlay!!.isEnabled = false
        chb_radio!!.isEnabled = false
        chb_classic!!.isEnabled = false
    }

    private fun stopPlaying() {
        playWhenReady = player!!.playWhenReady
        player!!.stop()

        buttonStopPlay!!.visibility = View.INVISIBLE
        buttonPlay!!.isEnabled = true
        buttonPlay!!.visibility = View.VISIBLE
        buttonStopPlay!!.isEnabled = false
        chb_classic!!.isEnabled = true
        chb_radio!!.isEnabled = true
    }

    override fun onAudioFocusChange(focusChange: Int) {
        when (focusChange) {
            AudioManager.AUDIOFOCUS_GAIN -> {
                // resume playback
                startPlaying()
            }

            AudioManager.AUDIOFOCUS_LOSS -> {
                // Lost focus for an unbounded amount of time: stop playback and release media player
                stopPlaying()
            }

            AudioManager.AUDIOFOCUS_LOSS_TRANSIENT -> {
                // Lost focus for a short time, but we have to stop
                // playback. We don't release the media player because playback
                // is likely to resume
                stopPlaying()
            }

            AudioManager.AUDIOFOCUS_LOSS_TRANSIENT_CAN_DUCK -> {
                // Lost focus for a short time, but it's ok to keep playing
                // at an attenuated level
                stopPlaying()
            }
        }
    }

    fun onCheckBoxClicked(view: View) {
        // Is the view now checked?
        val checked = (view as CheckBox).isChecked
        when (view.getId()) {
            R.id.chb_radio -> if (checked) {
                chb_classic!!.isChecked = false
            }
            R.id.chb_classic -> if (checked) {
                chb_radio!!.isChecked = false
            }
        }
    }

    private fun callStateListener() {
        // Get the telephony manager
        val telephonyManager = getSystemService(Context.TELEPHONY_SERVICE) as TelephonyManager
        //Starting listening for PhoneState changes
        val phoneStateListener: PhoneStateListener = object : PhoneStateListener() {
            override fun onCallStateChanged(state: Int, incomingNumber: String) {
                when (state) {
                    TelephonyManager.CALL_STATE_OFFHOOK, TelephonyManager.CALL_STATE_RINGING -> {
                        stopPlaying()
                    }
                    TelephonyManager.CALL_STATE_IDLE -> {
                        startPlaying()
                    }
                }
            }
        }
        // Register the listener with the telephony manager
        // Listen for changes to the device call state.
        telephonyManager.listen(phoneStateListener,
                PhoneStateListener.LISTEN_CALL_STATE)
    }

    companion object {
        private const val MY_PERMISSIONS_REQUEST_READ_PHONE_STATE = 1
        const val RADIO_STREAM = "http://116.202.18.122/radio/8010/radio.mp3"
        const val CLASSIC_STREAM = "http://116.202.18.122/radio/8000/radio.mp3?1584817375"
        const val Broadcast_PLAY_NEW_AUDIO = "org.allatra.radio.audioplayer.PlayNewAudio"
    }
}